### ISSUE TITLE
Kitchen conv(erge) gives AzureOperationError

### ISSUE THREAD

**Issue by iiro** [pendrica/kitchen-azurerm#24]
Hi,

I tried to follow all the setup instuctions, but I get this (in TK log file):

```
# cat .kitchen/logs/kitchen.log
I, [2016-03-18T23:16:00.544922 #82662]  INFO -- Kitchen: -----> Starting Kitchen (v1.6.0)
I, [2016-03-18T23:16:01.245511 #82662]  INFO -- Kitchen: -----> Creating <policy-web-apache2-ubuntu-1404>...
E, [2016-03-18T23:16:02.858463 #82662] ERROR -- Kitchen: ------Exception-------
E, [2016-03-18T23:16:02.858659 #82662] ERROR -- Kitchen: Class: Kitchen::ActionFailed
E, [2016-03-18T23:16:02.858815 #82662] ERROR -- Kitchen: Message: Failed to complete #create action: [MsRestAzure::AzureOperationError]
E, [2016-03-18T23:16:02.858883 #82662] ERROR -- Kitchen: ---Nested Exception---
E, [2016-03-18T23:16:02.858915 #82662] ERROR -- Kitchen: Class: MsRestAzure::AzureOperationError
E, [2016-03-18T23:16:02.858956 #82662] ERROR -- Kitchen: Message: MsRestAzure::AzureOperationError
E, [2016-03-18T23:16:02.858982 #82662] ERROR -- Kitchen: ------Backtrace-------
E, [2016-03-18T23:16:02.859005 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/azure_mgmt_resources-0.1.1/lib/azure_mgmt_resources/resource_groups.rb:305:in `block in begin_create_or_update'
E, [2016-03-18T23:16:02.859031 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/promise.rb:496:in `call'
E, [2016-03-18T23:16:02.859054 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/promise.rb:496:in `block in on_fulfill'
E, [2016-03-18T23:16:02.859078 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/safe_task_executor.rb:24:in `call'
E, [2016-03-18T23:16:02.859103 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/safe_task_executor.rb:24:in `block in execute'
E, [2016-03-18T23:16:02.859127 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/synchronization/mri_lockable_object.rb:38:in `block in synchronize'
E, [2016-03-18T23:16:02.859169 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/synchronization/mri_lockable_object.rb:38:in `synchronize'
E, [2016-03-18T23:16:02.859204 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/synchronization/mri_lockable_object.rb:38:in `synchronize'
E, [2016-03-18T23:16:02.859228 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/safe_task_executor.rb:19:in `execute'
E, [2016-03-18T23:16:02.859252 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/promise.rb:526:in `block in realize'
E, [2016-03-18T23:16:02.859276 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `call'
E, [2016-03-18T23:16:02.859299 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `run_task'
E, [2016-03-18T23:16:02.859323 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/ruby_thread_pool_executor.rb:337:in `block (3 levels) in create_worker'
E, [2016-03-18T23:16:02.859443 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `loop'
E, [2016-03-18T23:16:02.859514 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `block (2 levels) in create_worker'
E, [2016-03-18T23:16:02.859548 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `catch'
E, [2016-03-18T23:16:02.859574 #82662] ERROR -- Kitchen: /Users/iiro/.chefdk/gem/ruby/2.1.0/gems/concurrent-ruby-1.0.1/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `block in create_worker'
E, [2016-03-18T23:16:02.859599 #82662] ERROR -- Kitchen: ----------------------
```

Any ideas? Credential-file should be ok also.

I'm new to Azure to this can be easily an user error.

**Comment by stuartpreston** [comment]
Could you post your kitchen.yml file please? (feel free to take out the subscription ID and any sensitive values).

**Comment by stuartpreston** [comment]
I can repro this failure if I changed the driver_config location to a value that was not valid for my subscription.

```
driver_config:
  location: 'typod location'
```

Could that be the problem here?

**Comment by iiro** [comment]
Hi,

I tried to follow up your example, only changing the necessary ones - here's the Kitchen YAML:

```
---
driver:
  name: azurerm

driver_config:
  subscription_id: 'MY_SUBSCRIPTION_ID_MATCHING_IN_CREDENTIALS_FILE'
  location: 'West Europe'
  machine_size: 'Standard_D1'

transport:
  name: rsync
  username: ubuntu
  ssh_key: ~/.ssh/azure

provisioner:
  name: chef_zero

platforms:
  - name: ubuntu-14.04
    driver_config:
      image_urn: Canonical:UbuntuServer:14.04.3-LTS:latest
      vm_name: trusty-vm

suites:
  - name: policy_web_apache2
    provisioner:
      policyfile: policies/web_apache2.rb
    ## Example of how to use remote InSpec profiles.
    ## Requires Test-Kitchen 1.6.0+ and kitchen-inspec RubyGem version 0.12.4+
    #verifier:
    #  inspec_tests:
    #    - https://github.com/nathenharvey/tmp_compliance_profile
  - name: policy_web_nginx
    provisioner:
      policyfile: policies/web_nginx.rb
```

Can this be an error in my ~/.azure/credentials...? I tried to follow your example there as well...

**Comment by stuartpreston** [comment]
I used your config and was able to successfully create the resource group and provision the machine, so it feels like it could be something to do with the or the subscription producing a message that we don't deal with.

![capture](https://cloud.githubusercontent.com/assets/6384654/13898072/b951c8c0-edbe-11e5-8ef9-bad4cb12d9a9.PNG)

Are you able to use the azure-cli to create a group in case that comes up with an error too:

```azure group create kitchen-test "west europe"```

I've just pushed a v0.3.4 up to RubyGems that should show the real error coming back from Azure, please update your kitchen-azurerm gem and give it another go.

**Comment by iiro** [comment]
Thanks.

The "azure group create" command succeeds fine - but Test-Kitchen get's this error now:

´´´
       {"code"=>"AuthorizationFailed", "message"=>"The client '76c86e4c-e86f-4f2e-be07-c4eb1675910b' with object id '76c86e4c-e86f-4f2e-be07-c4eb1675910b' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourcegroups/write' over scope '/subscriptions/241973b3-397b-45f1-bfae-08ba9bdffa47/resourcegroups/kitchen-policy-web-apache2-ubuntu-1404-20160318T205243'."}
´´´

...so I have probably failed the permissions then...

**Comment by iiro** [comment]
Got this fixed now - by doing the "azure role assignment" with proper role.

For Test-Kitchen usage - what permissions would you recommend for...? Azure is not familiar to me - yet - but it would be good to know what would be the "sensible defaults" for TK usage - if you have any suggestions. :)

**Comment by iiro** [comment]
One thing more to ask - sorry for this... How to connect to the TK-created VM using an SSH key...? It seems that AzureRM did not create the instance with SSH key preinstalled - as I cannot access the machine (not using username azure or ubuntu)...

**Comment by stuartpreston** [comment]
Giving your Service Principal the Contributor access to your subscription is generally enough. 

SSH keys are only uploaded to the created machine when the transport is set to: ```ssh```

I'm not familiar with the rsync transport is there a particular reason you are using it?

**Comment by iiro** [comment]
Ok, I will switch to Contributor.

Ah - that's the logic here... At least with AWS instances (driver kitchen-ec2) the transport of Chef artifacts is superslow - as the SSH transport uses just scp to upload all files to provisioned node - and scp happens file by file; the sync takes minutes to happen... The RubyGem I've been using is kitchen-transport-rsync - it has worker really well.

I changed the transport to ssh now - to get things working - but I get another error now:

´´´
-----> Starting Kitchen (v1.6.0)
-----> Creating <policy-web-apache2-ubuntu-1404>...
       Creating Resource Group: kitchen-policy-web-apache2-ubuntu-1404-20160320T110721
       Creating Deployment: deploy-117f0c5dade37911
       Adding public key from /Users/iiro/.ssh/azure.pub to the deployment.
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: Failed to complete #create action: [399: unexpected token at '{
                "path": "[concat('/home/',parameters('adminUsername'),'/.ssh/authorized_keys')]",
                "keyData": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDyygBC4B/AyeH9UT4ejfV5vvpGbxxVxGeUpQPybI+3vwYAZ9y2DcrnQt0M/oJD8zAWTDTPvKcYzKDX9d1oKn9dPgeGpFZ3+BpSYF6kpSoik51enYbouMG156oaoFDlTht6ZP6lgvctUhTiKNLE5jgjTPsUU8BI1TVsuWMtbW0EvW5KsroHq77eB9AquQTXl126heBCn848l+F8Vi+H5GO8Q5Uorjwgkp5qLfwT5nkPtAsWUqA7M031Y81Pu3bBeYdKLdGln9KWflmW93DAiXFM5ausibe4hSQ66bEislEWKtQ0Qfxu03MyrT31OgOcom+23A4GiBDeI0hRScHFuCQb iiro@Ippes-Air.local
"
              }
            ]
          }
        }
']
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
´´´

The public key file is created with ssh-keygen without any tricks... I don't understand where the "unexpected token" is...

**Comment by stuartpreston** [comment]
Hmm neither do I. I'll give this another test later, have you tried moving azure and azure.pub out of the way? The driver will recreate new keys on the next create. If that works then I'd be interested in the differences between the keys.

**Comment by stuartpreston** [comment]
Actually is that a blank newline at the end of your public key file? I'm not sure any checking is done on that and newlines could break the JSON template behind the scenes I guess.

**Comment by iiro** [comment]
It was - superb. ssh-keygen (at least in OSX) seems to create the newline there by default...

Converge works now.

Btw - about the rsync transfer... If you run the converge with "-l debug" (and ssh as a transport) you can see TK to sync files with scp one file at a time. :)

**Comment by stuartpreston** [comment]
I've just pushed a couple of fixes as kitchen-azurerm 0.3.5 up to RubyGems. This removes the previous transport name restriction, and fixes the issue with the potential newline at the end of keys generated by ssh-keygen. Please give it a go with your rsync transport and let me know how you get on.

**Comment by iiro** [comment]
Works like a charm - with both the SSH pubkey end-of-line fix and the transport change. Superb!

I think we can for sure close this one :)
